### PR TITLE
feat: add blsgroup module

### PR DESCRIPTION
**Pull request by daeMOn63**
This module extends over the cosmos-sdk/x/group module and allows to:

- generate individual votes from BLS accounts
- aggregate all the individual votes and submit them all at once

Most of the security properties from the initial PRs should be there, except for the vote expiration time which I couldn't figure out how to add back, now that the proposal expiration belongs to the group policy rather than the proposal in itself. 

If needed, the previous group module work can be found on the https://github.com/fetchai/fetchd/tree/group_module branch.

I've added some basic test cases, mainly to ensure we have the proper setup and utilities to build more later. Ideally, we want to add more tests and validate all the security properties there.

### CODE REVIEW

**Comment by kitounliu** [comment]
About vote expiration time, this is mainly for preventing replaying attacks since aggregated vote cannot have a sequence number. If replacing vote is allowed, then without the timestamp, attacker could use an old vote to replace a new one.  

We only need to compare the vote expiration time with the block time to see if it is still valid.

**Comment by kitounliu** [comment]
I saw MsgVoteCmd() and GetVoteBasicCmd() are merged together as MsgVoteCmd(). But then we lose the capability for submitting vote directly onchain (without aggregation). If someone was later for the offchain vote aggregation, they should still be able to submit their vote directly onchain.

**Comment by kitounliu** [comment]
At the end of "func (k Keeper) VoteAgg", when adding the votes, I would just skip err of some votes since they are from different voters. A malicious voter could first vote online and then submit its vote again for aggregation to cause the whole aggregated vote to fail. The aggregated vote is expensive to process.

File: x/blsgroup/keeper/msg_server.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "load group policy")
+	}
+
+	electorate, err := k.groupKeeper.GroupInfo(goCtx, &group.QueryGroupInfoRequest{GroupId: policyInfo.Info.GroupId})
+	if err != nil {
+		return nil, err
+	}
+
+	groupMemberResp, err := k.groupKeeper.GroupMembers(goCtx, &group.QueryGroupMembersRequest{GroupId: electorate.Info.Id})
```
Comment (kitounliu) [inline x/blsgroup/keeper/msg_server.go:45]: The problem of  k.groupKeeper.GroupMembers is they use pagination and there is a default limit of 100. If we have more than 100 group members, we won't be able to get them all

**Comment by kitounliu** [comment]
POP_verify is missing...
When creating a group with bls members, we need to check their pops are verified. So we need to have a MsgCreateBlsGroup on server

**Comment by kitounliu** [comment]
> > I saw MsgVoteCmd() and GetVoteBasicCmd() are merged together as MsgVoteCmd(). But then we lose the capability for submitting vote directly onchain (without aggregation). If someone was later for the offchain vote aggregation or their votes were thrown away by a malicious combiner, they should still be able to submit their vote directly onchain.
> 
> Standalone votes can still be submitted with the standard group module (`tx group vote`): https://github.com/cosmos/cosmos-sdk/blob/5bde3686c4538ce53356af6e9fe40b34e4ce4a06/x/group/client/cli/tx.go#L681 ?

**Comment by kitounliu** [comment]
Sorry..I might've accidentally closed the pull request

**Comment by daeMOn63** [comment]
> POP_verify is missing... When creating a group with bls members, we need to check their pops are verified. So we need to have a MsgCreateBlsGroup on server

@kitounliu: on this, I'm not sure why we need to check for POP at group creation? As I understand it, for now, creating groups with rogue accounts shouldn't hurt, and the POP for each account is currently checked when submitting an aggregated signature (https://github.com/fetchai/fetchd/blob/bls_group_module/x/blsgroup/keeper/msg_server.go#L71-L74) - but I agree it's not super obvious, I'll add an explicit `if !bls12381.IsPopValid(acc)` check here to improve it. Unless there's an issue I didn't see yet on checking that at group creation.

**Comment by kitounliu** [comment]
The pop checking in VoteAgg is too late. Pop should be checked at entry point at group creation. If members are voting for multiple proposals, then all the aggregated votes will fail because one voter's pop is not checked.

**Comment by kitounliu** [comment]
Redundance and explicity are good for security. In addition, we need to consider usability and operational costs. We won't be able to predict all the security vulnerabilities since other parts of code might change in the future. What we can do is to make the security part more independent and robust.

File: x/blsgroup/keeper/msg_server.go (hunk 1 above)
Comment (daeMOn63) [reply x/blsgroup/keeper/msg_server.go:162]: Tracked in #242 - updating is easy but the SDK has a bug and fails to serve groups with many members.

**Comment by kitounliu** [comment]
Can we have CreateBlsGroup which can first call sdk's CreateGroup to create a group and then register the group as a bls group? And similarly the bls version of UpdateBlsGroupMembers/UpdateBlsGroupAdmin/UpdateBlsGroupMetadata/CreateBlsGroupWithPolicy/ UpdateBlsGroupPolicyMetadata/UpdateGroupPolicyDecisionPolicy. This will keep the bls group as registered when there is some change to the group. If someone calls a sdk group function on a bls group, the group will become unregistered.

**Comment by daeMOn63** [comment]
Yes, but I'd suggest doing this later and in other PRs. This one gives a first (unperfect) working version already big enough for a review.

**Review by kitounliu — APPROVED** [review]

**Pull request merged**